---
title: "Pre-processing ASVs with Phyloseq"
author: "Liangzi"
date: "2025-03-25"
output: html_document
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, 
                      fig.align = "center", 
                      # write figures to the figures folder
                      fig.path = "../figures/03_PreProcessing/")
```

# Set Environment 

## Load Packages 
```{r load-packages}
#install.packages("BiocManager")
#BiocManager::install("Biostrings")

# NOTE, you will need to install the following 2 packages to re-create the 
# negative control plots below!
#install.packages("ggpubr")
#install.packages("rstatix")

# Load packages with pacman
# Be sure that you load dada2 & Patchwork, which we didn't do in class on March 12th
pacman::p_load(devtools, phyloseq, dada2, patchwork, Biostrings, tidyverse,
               # Note to add ggpubr & rstatix!!
               ggpubr, rstatix, install = FALSE)
```

## Timing of this script
```{r rmd-start}
# What time did we start running this script? 
start_time <- Sys.time()
```

## 1. Load Data 

### 1a. Metadata 
```{r load-metadata}
# load in metadata
metadata_df <- 
  read_csv("data/metadata/SraRunTable.csv") %>%
  # Fix Column Name
  dplyr::rename("sample_names" = "Run") %>%
  # Add sample names also as a column 
  mutate(names = sample_names) 

# Inspect 
head(metadata_df)
dim(metadata_df)

# include dada2 output
load("data/track_read_counts.RData")

# Take a look
glimpse(track_counts_df)
dim(track_counts_df)

# Check filenames 
head(track_counts_df$sample_names)

# Fix sample names in track_reads 
track_counts_df$sample_names <- sapply(strsplit(track_counts_df$sample_names, "_"), `[`, 1)

# Intuition check 
head(track_counts_df$sample_names)

# What's different? 
setdiff(track_counts_df$sample_names, metadata_df$sample_names)

# Let's do a filtering join with left_join 
metadata_final_df <- 
  metadata_df %>%
  left_join(., track_counts_df, by = "sample_names") %>%
  # sample names to the rownames to merge into phyloseq
  column_to_rownames(var = "sample_names")

# Check 
dim(metadata_final_df)
```

### 1b. ASV Table 
```{r load-asv-table}
asv_df1 <- 
  read.delim(file = "data/ASV_table.tsv", sep = "\t",
           # add the column names and row names 
           header = TRUE, row.names = 1)

# Drop the selected columns
drop_cols <- c("SRR128733001", "SRR128733011", "SRR128733021", "SRR128733031",
               "SRR128733041", "SRR128733051", "SRR128733061", "SRR128733071", 
               "SRR128733081", "SRR128733091", "SRR128733101", "SRR128733111",
               "SRR128733121", "SRR128733131", "SRR128733141", "SRR128733151", 
               "SRR128733161", "SRR128733171", "SRR128733181", "SRR128733191", 
               "SRR128733201", "SRR128733211", "SRR128733221", "SRR128733231",
               "SRR128733241", "SRR128733251", "SRR128733261", "SRR128733271",
               "SRR128733281")

asv_df <- asv_df1 %>%
  select(-all_of(drop_cols))

# Inspect 
asv_df[1:3, 1:3]

# fix Column names 
## Remove the X: denote at the beginning "^"
colnames(asv_df) <- sub(pattern = "^X", replacement = "" , colnames(asv_df))
## Replace the . with a -: "\\." since . is a special character in regex
colnames(asv_df) <- gsub(pattern = "\\.", replacement = "-", colnames(asv_df))

# Final inspection 
head(colnames(asv_df))
asv_df[1:3, 1:3]
```

### 1c. Taxonomy Table
```{r load-tax-table}
tax_df <- 
  read.delim("data/ASV_taxonomy.tsv", sep = "\t",
           header = TRUE, row.names = 1) 

# Inspect
dim(tax_df)
dim(asv_df)

# Add check 
stopifnot(rownames(asv_df) == rownames(tax_df))
```

# 2. Handoff to phyloseq 
```{r phyloseq-handoff}
raw_physeq <- 
  phyloseq(otu_table(asv_df, taxa_are_rows = TRUE),
         sample_data(metadata_final_df),
         tax_table(as.matrix(tax_df)))

# Check out 
raw_physeq

# save 
save(raw_physeq, file = "data/03_PreProcessing/raw_physeq.RData")
```

# Clean up the data! 

## 3. Remove chloroplasts
```{r rm-chloro}
noChloros_physeq <- 
  raw_physeq %>%
  subset_taxa(Order != "Chloroplast" | is.na(Order))

# How many ASVs were chloroplasts? 
numChloros_ASVs <- ntaxa(raw_physeq) - ntaxa(noChloros_physeq)

numChloros_ASVs # There were 3 ASVs that were chloroplasts.
```

## 4. Remove Mitochondria 
```{r rm-mitos}
noChlorosMitos_physeq <-  
  noChloros_physeq %>%
  subset_taxa(Family != "Mitochondria" | is.na(Family)) 

# How many ASVs were mitochondria? 
numMitos_ASVs <- ntaxa(noChloros_physeq) - ntaxa(noChlorosMitos_physeq)

numMitos_ASVs # There were 9 ASVs that were mitochondria.
```

**INTERPRETATION #1: How many chloroplast and/or mitochondrial ASVs in your dataset? Does this seem reasonable?**
There were 3 ASVs that were chloroplasts and 9 ASVs that were mitochondria.
This seem reasonable since the dataset is not plant-heavy, and maybe the primer selection minimized chloroplast and mitochondrial amplification.

# Evaluate and remove the control samples 

## 5. Negative Controls 
```{r neg-control-physeq, fig.width=7, fig.height=3}
# # List all sample names to check if negative control samples exist
sample_names(noChlorosMitos_physeq)
```

**INTERPRETATION #2: Did your dataset have any negative controls? If so, how many ASVs did you remove? Does that seem reasonable to you?**
In this paper, the authors did not analyze the microbial community for control groups. Negative control samples don't exist in the 'noChlorosMitos_physeq'.

## 6. Positive Controls 

**INTERPRETATION #3: Did your dataset have a mock community? If so, what can you say about the accuracy of your sequencing run?**
In this paper, the authors did not analyze the microbial community for control groups.There is no mock community.

# 7. Reverse Complements 
```{r reverse-complements}
# Pull out Taxa Table
tax_table <- 
  noChlorosMitos_physeq %>%
  tax_table() %>% 
  data.frame()

# Grab ASV Sequences and Names
asv_names <- tax_table$ASV
asv_seqs <- tax_table$ASVseqs

# Create empty comparison matrix
asv_rev_comp <- matrix(nrow=length(asv_seqs), ncol = length(asv_seqs)) 

# Fix the names in the rows and columns 
rownames(asv_rev_comp) <- asv_names
colnames(asv_rev_comp) <- asv_names

# Convert sequences to Biostrings
asv_dna <- DNAStringSet(asv_seqs) 

# Inspect 
head(asv_dna)

# Construct reverse complement
asv_rev_dna <- reverseComplement(asv_dna) 

# Now loop through every asv sequence to check 
for(i in 1:length(asv_seqs)){ # For each asv sequence...
  match_vec <- asv_dna[i] == asv_rev_dna # Compare it to the reverse complement of every other sequence...
  asv_rev_comp[,i] <- match_vec # Write that logical vector as a new column 
}

# Find how many TRUEs (matches) we have, divide by 2 because each pair occurs twice
cat("For", sum(asv_rev_comp) / 2,"ASVs, the reverse complement will need to be removed") 
```

**INTERPRETATION #4: Do you have any reverse complements in your dataset?**
Output: For 0 ASVs, the reverse complement will need to be removed.
Therefore, the dataset does not contain any reverse complement ASVs that need to be removed.

# 8. Sequencing Depth 
```{r seq-depth, fig.height=3, fig.width=8}
# The current data object
noChlorosMitos_physeq

# What is the library size/sequencing depth for each sample? 
seqSums_df <- 
  noChlorosMitos_physeq %>%
  otu_table() %>%
  # Sum each sample column 
  colSums() %>%
  data.frame() %>%
  rownames_to_column(var = "names") %>%
  left_join(., metadata_final_df, by = "names") 

# Rename second column 
colnames(seqSums_df)[2] <- "TotalSeqs"

# check
dim(seqSums_df)
head(seqSums_df)

# Show the depth of samples 
seqSums_df %>%
  dplyr::select(names, TotalSeqs) %>%
  arrange(TotalSeqs) %>%
  head()

# plot it as a bar plot 
numSeq_bar_plot <- 
  seqSums_df %>%
  ggplot(aes(x=reorder(names, TotalSeqs), y = TotalSeqs,
             fill = geo_loc_name)) + 
  geom_bar(stat = "identity") + 
  theme(axis.text.x = element_blank()) + 
  labs(y = "Read Depth", x = "Sample") + 
  theme(legend.position = "none")

# histogram
numSeq_hist_plot <- 
  seqSums_df %>%
  ggplot(aes(x= TotalSeqs, fill = geo_loc_name)) + 
  geom_histogram(color = "black") + 
  labs(y = "# of Samples", x = "Read Depth") + 
  theme(legend.position = "bottom")

# Density plot 
numSeq_density_plot <- 
  seqSums_df %>%
  ggplot(aes(TotalSeqs, fill = geo_loc_name)) +
  geom_density(alpha = 0.5) + 
  labs(x = "Read Depth") + 
  theme(legend.position = "none")

# Put it all together 
numSeq_bar_plot + numSeq_hist_plot + numSeq_density_plot + 
  plot_annotation(tag_levels = "A") 
```
**INTERPRETATION #5: What can you conclude about your sequencing depths? What are your sample distributions? Do you have “enough” reads? Are there any samples that should be removed at this step?**

*Conclusion:*
Panel A: Bar Plot of Read Depth per Sample
The sequencing depth varies significantly across samples.
Some samples have much lower read counts than others.
If there are samples with very low sequencing depths, they may need to be removed to maintain data quality.

Panel B: Histogram of Read Depth Distribution
The distribution appears right-skewed, meaning most samples have relatively high sequencing depths, but a few have much lower counts.
The presence of samples with very low read depths (close to zero) may indicate sequencing failures or contamination.
If a threshold was set, samples below this should be considered for removal.

Panel C: Density Plot of Read Depth
The distribution is bimodal, suggesting that there may be two groups of samples:
1) A group with lower read depths
2) A group with higher read depths
This could indicate a mix of different sequencing qualities or batches.

*Sample Distributions:*
There is a bimodal distribution, suggesting variability in sequencing depth between samples.
This could be due to differences in DNA extraction, library preparation, or sequencing efficiency.

*“enough” reads?*
It seems like I have enough reads. 

*Any samples that should be removed?*
Yes, there are some samples that should be removed. 
The outliers on the left side of the bar plot and histogram may be unreliable and should be excluded. In order to ensure even sequencing coverage across samples, I need to remove samples with a zero read depth. 

## Remove samples with few reads
```{r rm-samps}
# What's the min seq depth? 
min(sample_sums(noChlorosMitos_physeq))

samples_to_remove <- c("SRR12873307", "SRR12873308", "SRR12873309", "SRR12873310",
                       "SRR12873311", "SRR12873312", "SRR12873313", "SRR12873314", "SRR12873317",
                       "SRR12873315")
raw_preprocessed_physeq <- 
  subset_samples(noChlorosMitos_physeq, !(sample_names(noChlorosMitos_physeq) %in% samples_to_remove))

# What's the new min seq depth?
min(sample_sums(raw_preprocessed_physeq))

# Final check of the sequencing depth for now
# Here, we could remove more samples if we needed to
seqSums_df %>%
  dplyr::select(names, TotalSeqs) %>%
  arrange(TotalSeqs) %>%
  head(n = 10)
```

# 9. Save Output 
```{r save-physeq}
save(raw_preprocessed_physeq, file = "data/03_PreProcessing/raw_preprocessed_physeq.RData")
```

# Final info for Reproducibility 

## Check Render Time
```{r stop-time}
# Take the time now that we are at the end of the script
end_time <- Sys.time()
end_time 

# Echo the elapsed time
elapsed_time <- round((end_time - start_time), 3)
elapsed_time
```

## Session Information
```{r session-info}
# Ensure reproducibility with package version information
devtools::session_info()
```


